src/persist/src/indexed/runtime.rs

```diff
@@ -0,0 +1,233 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! WIP
+
+#![allow(missing_docs)]
+
+use std::sync::mpsc::{Receiver, RecvError, SendError};
+use std::sync::{mpsc, Arc, Mutex};
+use std::thread::{self, JoinHandle};
+
+use log;
+
+use crate::error::Error;
+use crate::indexed::{Indexed, IndexedSnapshot};
+use crate::storage::{Blob, Buffer};
+use crate::Id;
+
+enum Cmd {
+    Register(Id, Box<dyn FnOnce(Result<(), Error>) + Send + 'static>),
+    Write(
+        Id,
+        Vec<((String, String), u64, isize)>,
+        Box<dyn FnOnce(Result<(), Error>) + Send + 'static>,
+    ),
+    Seal(Id, u64, Box<dyn FnOnce(Result<(), Error>) + Send + 'static>),
+    Snapshot(
+        Id,
+        Box<dyn FnOnce(Result<IndexedSnapshot, Error>) + Send + 'static>,
+    ),
+    Stop,
+}
+
+#[derive(Clone)]
+pub struct RuntimeCore {
+    handle: Arc<Mutex<Option<JoinHandle<()>>>>,
+    tx: mpsc::Sender<Cmd>,
+}
+
+impl RuntimeCore {
+    fn send(&self, cmd: Cmd) {
+        if let Err(SendError(_)) = self.tx.send(cmd) {
+            // According to the docs, a SendError can only happen if the
+            // receiver has hung up, which only happens if the thread has
+            // exited. The thread only exits if we send it a Cmd::Stop or it
+            // panics. Cmd::Stop is only sent by RuntimeCore::stop, which
+            // consumes self, so it must be a panic.
+            log::error!("unable to send command; the runtime thread must have panic'd");
+        }
+    }
+
+    fn stop(self) {
+        if let Some(handle) = self.handle.lock().expect("WIP").take() {
+            self.send(Cmd::Stop);
+            if let Err(_) = handle.join() {
+                // If the thread panic'd, then by definition it has been
+                // stopped, so we can return an ok. This is surprising, though,
+                // so log a message. Unfortunately, there isn't really a way to
+                // put the panic message in this log.
+                log::error!("persist runtime thread panic'd");
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Runtime {
+    core: Option<RuntimeCore>,
+}
+
+impl Runtime {
+    pub fn start<U: Buffer + Send + 'static, L: Blob + Send + 'static>(
+        indexed: Indexed<U, L>,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel();
+        let handle = thread::spawn(move || {
+            let mut l = RuntimeLoop { indexed, rx };
+            while l.work() {}
+        });
+        let handle = Arc::new(Mutex::new(Some(handle)));
+        let core = RuntimeCore { handle, tx };
+        Runtime { core: Some(core) }
+    }
+
+    pub fn register(&self, id: Id) -> Result<(), Error> {
+        let core = self.get_core("register")?;
+
+        let (tx, rx) = mpsc::channel();
```

Comment:
Did you want to eventually have something like a `Response` enum and having `RuntimeLoop` store a `tx` to send different kinds of responses back? I think the current approach gives some tighter guarantees about the types of potential responses 

Thinking a little more carefully I suppose having a new channel for each operation also protects against bad interleavings (two threads call `write_sync` at the same time, but then thread B gets the response for thread A's write).

Reply:
the new enum looks great to me